clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,216 @@
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel4>([]() {});
+}
+
+void callkernel4() {
+  kernel4func([]() {}); // #kernel4func_call
+}
+
+template <template <typename> typename Outer, typename Inner>
+struct S {
+  void operator()() const;
+};
+
+template <typename Ty>
+struct Tangerine {};
+
+template <typename Func>
+void kernel13_14func(const Func &F) {
+  kernel_single_task<class kernel13>(F);
+  kernel_single_task<class kernel14>(F);
+  // Using the same functor twice should be fine
+}
+
+// kernel13 and kernel14 - expect no errors
+// Pass the same lambda to two kernels
+// Test that passing the same lambda to two kernels does not cause an error
+// because the kernel uses do not interfere with each other or invalidate
+// the stable name in any way
+void callkernel13_14() {
+  kernel13_14func([]() {});
+}
+
+template <typename T>
+static constexpr const char *output1 = __builtin_unique_stable_name(T);
+
+#define MACRO12()                \
+  auto l12 = []() { return 1; }; \
+  constexpr const char *l1_output = __builtin_unique_stable_name(l12); // #USN_l12
+
+int main() {
+
+  // kernel1 - expect no error
+  // Evaluate the builtin in a constant context, then call the kernel with
+  // that same type
+  // Tests that passing the lambda to the unique stable name builtin and
+  // then using the lambda in a way that does not  contribute to the kernel
+  // name does not cause an error because the  stable name is not invalidated
+  // in this situation.
+  auto l1 = []() {};
+  constexpr const char *l1_output = __builtin_unique_stable_name(l1);
+  kernel_single_task<class kernel1>(
+      [=]() { l1(); }); // Used in the kernel, but not the kernel name itself
+
+  // kernel2 - expect error
+  // Call to the builtin on a lambda, then using the lambda for a kernel name
+  // Test that passing the lambda to the USN builtin and then using the same
+  // lambda in the naming of a kernel causes a diagnostic on the kernel use due
+  // to the change in results to the stable name.
+  auto l2 = []() { return 1; };
+  constexpr const char *l2_output = __builtin_unique_stable_name(l2); // #USN_L2
+  // expected-error@#kernelSingleTask{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@#USN_L2{{'__builtin_unique_stable_name' evaluated here}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel2>(l2); // Used in the kernel name after builtin
+
+  // kernel7 - expect error
+  // Same as kernel6, except make l6 part of naming kernel
+  // Test that passing a lambda to the unique stable name builtin and then
+  // passing a second lambda to the kernel throws an error because the first
+  // lambda is included in the signature of the second lambda, hence attempting
+  // to change the mangling of the kernel
```

Comment:
```suggestion
  // Same as kernel6 (below), except make the lambda part of naming the kernel.
  // Tests that passing a lambda to the unique stable name builtin and then
  // passing a second lambda to the kernel throws an error because the first
  // lambda is included in the signature of the second lambda, hence it changes 
  // the mangling of the kernel.
```